common/network-common/src/main/java/org/apache/spark/network/crypto/GcmTransportCipher.java

```diff
@@ -0,0 +1,356 @@
+                    throw new RuntimeException(e);
+                }
+                ciphertextBuffer.flip();
+                int outputRemaining = ciphertextBuffer.remaining();
+                while (ciphertextBuffer.hasRemaining()) {
+                    target.write(ciphertextBuffer);
+                }
+                transferredThisCall += outputRemaining;
+                transferred += outputRemaining;
+            }
+            return transferredThisCall;
+        }
+
+        private long getReadableBytes() {
+            if (plaintextMessage instanceof ByteBuf byteBuf) {
+                return byteBuf.readableBytes();
+            } else if (plaintextMessage instanceof AbstractFileRegion fileRegion) {
+                return fileRegion.count() - fileRegion.transferred();
+            } else {
+                throw new IllegalArgumentException("Unsupported message type: " +
+                        plaintextMessage.getClass().getName());
+            }
+        }
+
+        @Override
+        protected void deallocate() {
+            if (plaintextMessage instanceof ReferenceCounted referenceCounted) {
+                referenceCounted.release();
+            }
+            plaintextBuffer.clear();
+            ciphertextBuffer.clear();
+        }
+    }
+
+    @VisibleForTesting
+    class DecryptionHandler extends ChannelInboundHandlerAdapter {
+        private final ByteBuffer ciphertextBuffer;
+        private final ByteBuffer plaintextBuffer;
+        private final AesGcmHkdfStreaming aesGcmHkdfStreaming;
+        private final StreamSegmentDecrypter decrypter;
+        private boolean decrypterInit = false;
+        private int segmentNumber = 0;
+        private long expectedLength = -1;
+        private long ciphertextRead = 0;
+
+        DecryptionHandler() throws GeneralSecurityException {
+            aesGcmHkdfStreaming = new AesGcmHkdfStreaming(
+                    aesKey.getEncoded(),
+                    "HmacSha256",
+                    aesKey.getEncoded().length,
+                    CIPHERTEXT_BUFFER_SIZE,
+                    0);
+            plaintextBuffer =
+                    ByteBuffer.allocate(aesGcmHkdfStreaming.getPlaintextSegmentSize());
+            ciphertextBuffer =
+                    ByteBuffer.allocate(aesGcmHkdfStreaming.getCiphertextSegmentSize());
+            decrypter = aesGcmHkdfStreaming.newStreamSegmentDecrypter();
+        }
+
+        @Override
+        public void channelRead(ChannelHandlerContext ctx, Object ciphertextMessage)
+                throws GeneralSecurityException {
+            Preconditions.checkArgument(ciphertextMessage instanceof ByteBuf,
+                    "Unrecognized message type: %s",
+                    ciphertextMessage.getClass().getName());
+            ByteBuf ciphertextNettyBuf = (ByteBuf) ciphertextMessage;
+            // The format of the output is:
+            // [8 byte length][Internal IV and header][Ciphertext][Auth Tag]
+            try {
```

Comment:
In this method, we cannot make assumptions about how much data is available to read from the incoming `ByteBuf` - when reading a segment, or even when we are reading the header.

These cases are currently not handled.

Reply:
Why is `readableBytes()` not the actual readable bytes? The `ByteBuf.readBytes(ByteBuffer)` doesn't return the number of bytes read.

Reply:
The `CtrTransportCipher` implementation also relies on the input buffer's `readableBytes` being accurate:
``` 
byte[] decryptedData = new byte[buffer.readableBytes()];
byteChannel.feedData(buffer);
```

Reply:
`readableBytes` would be accurate - but we cant assume `readableBytes` is >= what we need.
I would suggest to split this method more explicitly, something like:
```
if (! headerRead) {
  if (! tryReadingHeader(ciphertextNettyBuf )) return ;
  headerRead = true;
}

if (! decrypterInit) {
  if (! tryDecrypterInit(ciphertextNettyBuf )) return ;
  decrypterInit= true;
}

// All initialization has been done already - decrypt the incoming message
while (ciphertextNettyBuf.readableBytes() > 0) {
  // copy to ciphertextBuffer, break if ciphertextBuffer is not fuil
  // when ciphertextBuffer is full (or last chunk handling), decrypt + fire event.
}

```